### PR TITLE
fix hashing regression.

### LINKED ISSUE #50628: Wrong hashes for some `Rational`s
On Julia v1.10.0-alpha1 and master, some `Rational`s which are equal have different hashes:
```julia
julia> hash(5//2) # same hash as on older Julia versions (≥ v1.6)
0x16b7f40fdbb2c57d

julia> hash(Int128(5)//2) # the same hash is also obtained for UInt128(5)//2 and BigInt(5)//2
0xb795033f6f2a0674
```

### PR DESCRIPTION
**Pull request by oscardssmith**
This fixes 2 bugs introduced by #49996 and #50041.
Closes #50628.

### CODE REVIEW

File: base/float.jl [hunk 1]
```diff
@@ -688,22 +688,19 @@ function hash(x::Real, h::UInt)
...
     den >>= den_z
     pow += num_z - den_z
-
     # handle values representable as Int64, UInt64, Float64
     if den == 1
         left = top_set_bit(abs(num)) + pow
-        right = pow + den_z
-        if -1074 <= right
-            if 0 <= right
+        if -1074 <= pow
```
Comment (gbaraldi) [inline base/float.jl:694]: Could you add a quick comment as to how we do this hashing?

File: base/float.jl [hunk 2]
```diff
@@ -688,22 +688,25 @@ function hash(x::Real, h::UInt)
     den_z = trailing_zeros(den)
     den >>= den_z
     pow += num_z - den_z
-
-    # handle values representable as Int64, UInt64, Float64
+    # If the real is an Int64, UInt64, or Float64, hash as those types.
+    # To be an Integer the denominator must be 1 and the power must be positive.
+    # To be a Float64
```
Comment (KristofferC) [inline base/float.jl:693]: Something missing here?

Comment (oscardssmith) [reply base/float.jl]: oops. That got moved to the comment on line 704

File: base/float.jl [hunk 3]
```diff
@@ -688,22 +688,24 @@ function hash(x::Real, h::UInt)
     den_z = trailing_zeros(den)
     den >>= den_z
     pow += num_z - den_z
-
-    # handle values representable as Int64, UInt64, Float64
+    # If the real is an Int64, UInt64, or Float64, hash as those types.
+    # To be an Integer the denominator must be 1 and the power must be positive.
```
Comment (LilithHafner) [inline base/float.jl:692]:
```suggestion
    # To be an Integer the denominator must be 1 and the power must be non-negative.
```

File: base/float.jl [hunk 4]
```diff
@@ -688,22 +688,24 @@ function hash(x::Real, h::UInt)
...
     if den == 1
+        # left = ceil(log2(num)*2^pow)
```
Comment (LilithHafner) [inline base/float.jl:694]:
```suggestion
        # left = ceil(log2(num*2^pow))
```
idk how much value this comment adds, but if we keep it, let's fix this

File: base/float.jl [hunk 5]
```diff
@@ -688,22 +688,24 @@ function hash(x::Real, h::UInt)
...
+    # To be an Integer the denominator must be 1 and the power must be non-negative.
     if den == 1
+        # left = ceil(log2(num*2^pow))
         left = top_set_bit(abs(num)) + pow
-        right = pow + den_z
-        if -1074 <= right
-            if 0 <= right
+        # 2^-1074 is the minimum Float64 so if the power is smaller, not a Float64
+        if -1074 <= pow
+            if pow >= 0 # if pow is negative, it isn't an integer
```
Comment (LilithHafner) [inline base/float.jl:698]:
```suggestion
            if 0 <= pow # if pow is non-negative, it is an integer
```
I think this wording is more natural

**Comment by oscardssmith** [comment]
Thanks @LilithHafner for the doc review. Good to get a second set of eyes on this.

**Review by LilithHafner — APPROVED** [review]
Functionality LGTM if tests pass.
Not sure what we were doing with `right`, it seems silly to me now.
Thanks for adding comments, too.

**Comment by oscardssmith** [comment]
we should probably add tests for rational numbers that are subnormal floats and typemax float

**Comment by oscardssmith** [comment]
The subnormal cases work. I think this is good to merge.

**Comment by vtjnash** [comment]
Please try to make sure you are using squash-merge when appropriate. It is hard to backport messy commits like this, odd to see that you are adding typo's to the language according to the commit logs, and not great for bisecting.

**Comment by KristofferC** [comment]
Also https://github.com/JuliaLang/julia/pull/50525#issuecomment-1640251473.

**Pull request merged**